File: c_src/2.c
struct ListNode *addTwoNumbers(struct ListNode *l1, struct ListNode *l2) {
    struct ListNode *dummyHead = (struct ListNode*)calloc(1, sizeof(struct ListNode));
    dummyHead->val = 0;
    int carry = 0, x, y, sum;
    struct ListNode *p = l1, *q = l2, *curr = dummyHead;
    while (p || q) {
        x = p ? p->val : 0;
        y = q ? q->val : 0;
        sum = x + y + carry;
        carry = sum / 10;
        struct ListNode *next = (struct ListNode*)calloc(1, sizeof(struct ListNode));
        next->val = sum % 10;
        curr->next = next;
        curr = curr->next;
        if (p) {
            p = p->next;
        }
        if (q) {
            q = q->next;
        }
    }
    if (carry > 0) {
        struct ListNode *next = (struct ListNode*)calloc(1, sizeof(struct ListNode));;
        next->val = carry;
        curr->next = next;
    }
    return dummyHead->next;
}

File: java_src/2.java
/**
 * Definition for singly-linked list.
 */

 public class ListNode {
     int val;
     ListNode next;
     ListNode(int x) { val = x; }
 }

 class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
         ListNode dummyHead = new ListNode(0);
         ListNode p = l1, q = l2, cur = dummyHead;

         int x, y, sum, carry = 0;
         while (p != null || q != null) {
             x = p == null ? 0 : p.val;
             y = q == null ? 0 : q.val;
             sum = x + y + carry;
             carry = sum / 10;
             cur.next = new ListNode(sum % 10);
             cur = cur.next;
             if (p != null) {
                 p = p.next;
             }
             if (q != null) {
                 q = q.next;
             }
         }
         if (carry > 0) {
             cur.next = new ListNode(carry);
         }
         return dummyHead.next;
     }
 }